File: problems/贪心/455.分发饼干.py
#!/usr/bin/env python3
"""
455. 分发饼干
算法: 贪心
时间复杂度: O(n log n)
空间复杂度: O(1)
"""
def find_content_children(g: list[int], s: list[int]) -> int:
    g.sort()
    s.sort()
    i = j = 0
    while i < len(g) and j < len(s):
        if s[j] >= g[i]:
            i += 1
        j += 1
    return i


if __name__ == "__main__":
    print(find_content_children([1, 2, 3], [1, 1]))  # 预期: 1
    print(find_content_children([1, 2], [1, 2, 3]))  # 预期: 2
    print(find_content_children([10, 9, 8, 7], [5, 6, 7, 8]))  # 预期: 2

File: problems/贪心/455.分发饼干.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int findContentChildren(vector<int>& g, vector<int>& s) {
    sort(g.begin(), g.end());
    sort(s.begin(), s.end());
    int i = 0, j = 0;
    while (i < g.size() && j < s.size()) {
        if (s[j] >= g[i]) i++;
        j++;
    }
    return i;
}

int main() {
    // 测试用例
    vector<int> g1 = {1, 2, 3}, s1 = {1, 1};
    cout << findContentChildren(g1, s1) << endl;  // 预期: 1

    vector<int> g2 = {1, 2}, s2 = {1, 2, 3};
    cout << findContentChildren(g2, s2) << endl;  // 预期: 2

    vector<int> g3 = {10, 9, 8, 7}, s3 = {5, 6, 7, 8};
    cout << findContentChildren(g3, s3) << endl;  // 预期: 2

    return 0;
}
